### DISCUSSION TITLE
QuestPDF Companion Development

### DISCUSSION THREAD

**Discussion by MarcinZiabek** [QuestPDF/QuestPDF#621] [Ideas]
## Introduction

The QuestPDF Previewer application is an innovative tool designed to showcase a real-time, dynamic preview of a document presently under development, utilizing a hot-reload mechanism without requiring code recompilation. This concept, originally proposed and initiated by @bennetbo, has proven immensely beneficial and worth further enhancements. 

## Planned capabilities

The QuestPDF Companion is envisioned as an evolved successor to the Previewer application, offering a set of additional features and functionalities that include:

- Identifying visible layout structures upon clicking any area of the document,
- Displaying attributes related to the selected layout element (size, position, layout status, properties),
- Presenting the document's structure in an intuitive, tree-like view,
- Enhancing the visualization of layout overflow exceptions through a tree structure rather than text,
- Detecting potential performance bottlenecks and optimization opportunities,
- Option to run the preview without hot-reload functionality by utilizing pooling,
- Advanced compatibility with diverse library versions and communication protocols,
- Support for light and dark modes.

## Current Status

This next-generation previewer project has been in active development for several months. While a considerable portion of the application has reached completion and demonstrates stability, I've encountered some peculiar issues while using Avalonia 10. These challenges could be attributed to my limited familiarity with the library, possible incorrect usage, or pre-existing bugs and issues. My daily work involves modern JavaScript technologies such as Vue.js, making the experience with Avalonia seem unproductive and unnecessarily complex.

## Roadmap

In light of the frequent problems, while developing the new version's implementation, I have chosen to reassess my strategy. Currently, I am contemplating two potential paths:

### Investigate Avalonia 11

This recent release appears to demonstrate superior quality. Many of my existing difficulties might be readily resolved, improving my satisfaction with this library as a solid foundation and enabling me to keep up the progress achieved over the past few months.

### Investigate Flutter

This library, written in Dart and similarly built upon the Skia library, is architecturally closer to modern frameworks, potentially diminishing future development costs. Although my understanding of this library is still at a foundational level, I am notably impressed by it so far. Nevertheless, this route does present a handful of potential issues, for instance, the inability to release the application as a dotnet tool.

## Afterword 

Over the upcoming weeks, I will update you on the progress regularly and hopefully gather valuable feedback.

**Comment by MarcinZiabek** [comment]
I've upgraded the QuestPDF Companion project to Avalonia 11. While a few adjustments were needed, the official guide is detailed and accurate. I'm genuinely impressed with the new version's stability. Those quirks and glitches I ran into earlier are fixed now. Turns out they were library bugs, not my lack of understanding, which gives me more confidence. 

Kudos to the Avalonia team; the new version is very imposing!

I am planning to share a few screenshots soon, followed by making the code public. While it's still a work in progress, the current version gives a glimpse into the direction I'm taking.

I'm at a crossroads about continuing with Avalonia or switching gears to Flutter. However, it's less about the platform's evolution and more about the overall development paradigm. With rumors of Flutter possibly transitioning to the Propeler engine and away from Skia, I fear future integration challenges between QuestPDF and the Companion tool.

This effort is not my highest priority at this moment, so there is still time for experimentation 😀

**Comment by MarcinZiabek** [comment]
## Introduction

Several months ago, QuestPDF underwent a significant update, transitioning from SkiaSharp to a custom-built Skia dependency. This remarkable transition was a testament to the power of our community, whose invaluable feedback, testing, and suggestions made it all possible. Building on this positive experience, we're excited to initiate a similar collaborative process for designing the next generation of the QuestPDF Companion application.

## Current progress

While this post doesn't detail the entire roadmap, we want to share that development has been ongoing for months and is now close to the public release. Recent QuestPDF updates have introduced key changes and features to facilitate this effort, which is now our primary focus.

The QuestPDF Companion will be an essential tool for all QuestPDF developers. We've previously shared ideas for advanced tooling and will regularly post screenshots and descriptions of specific features. We encourage you to share your thoughts and suggestions as we iterate on the design.

## Technicals

From a technical standpoint, the application will be built using Flutter and support Windows, MacOS, and Linux platforms. It will be directly downloaded from the project homepage rather than as a NuGet package.

## First screenshot

We're excited to share a screenshot of the application's home screen. Say goodbye to the empty gray QuestPDF Previewer!

<img width="512" alt="image" src="https://github.com/user-attachments/assets/95683375-7790-4602-be7f-082c6bf22320">


<img width="512" alt="image" src="https://github.com/user-attachments/assets/401bf33c-de23-424e-807d-9e4e7a624847">

**Comment by MarcinZiabek** [comment]
The next area of improvement is the design of the generic exception view. Previously, such exceptions were handled and formatted using a custom-formatted document. Now, the tool displays proper exceptions with a listed stack trace and clickable links.

<img width="992" alt="image" src="https://github.com/user-attachments/assets/4590172f-df05-40cb-b639-718a74bca90e">

<img width="992" alt="image" src="https://github.com/user-attachments/assets/9bc6a9cd-e54c-4f37-9008-4800fe7b4150">

You may also find two additional icons in the top right corner: one indicates the current license configured in the code calling previewer, and the second provides quick access to settings.

You can select which IDE should be used to open all clickable links.

<img width="838" alt="image" src="https://github.com/user-attachments/assets/eb5f1a02-7589-4dca-88f0-a5cf18185961">

**Reply by kozw** [reply]
Flutter FTW! 😍

**Comment by MarcinZiabek** [comment]
## Document Preview View

In this development update, I am excited to introduce the primary Document Preview View in our application. This view is designed to streamline your workflow and enhance your document management experience. It consists of two main components:
- Document Structure Tree (left-hand side)
- Document Preview (right-hand side)

### Document Structure Tree

The Document Structure Tree presents the layout of your document in a hierarchical tree format, similar to the Fluent API structure. Here are some key features:

- Condensed Indentation: The structure tree uses a condensed indentation style, making it easier to navigate through complex documents.
- Distinctive Icons: Specific elements are marked with unique icons, allowing you to quickly identify and distinguish between different types of content.
- Visibility Indicators: Elements that are not currently visible on the screen are dimmed, helping you maintain your orientation within extensive documents. This feature uses bounding boxes to ensure accuracy.
- Semantic Highlighting: Important semantic elements such as document parts, sections, and other critical components are highlighted for better visibility and understanding.
- Dynamic Naming: Some elements may appear under different names to improve clarity. For instance, a section is displayed using its actual name for easy identification.
- Additional Layout Insights: The library captures layout structures like class and method names in the source code, enabling you to quickly locate the code responsible for specific content.
- Text Hints: The content of text elements is displayed as hints, providing additional context and information at a glance.

### Document Preview
The Document Preview offers a comprehensive view of your document, similar to the previous iteration of our application. We are planning to introduce a range of interactive, navigational, and measurement features to enhance your experience further. Stay tuned for more exciting updates in future releases.

<img width="1222" alt="image" src="https://github.com/user-attachments/assets/b11f08eb-fc22-4ad3-88b2-9d17acdfb52f">
<img width="1222" alt="image" src="https://github.com/user-attachments/assets/357ba16d-40e4-4425-aea8-96e850a0abf7">

**Reply by yudielcurbelo** [reply]
It looks great. When we would be able to try it out?

**Reply by MarcinZiabek** [reply]
Thank you 😄 

The plan is as follows:
1) Release the first version at the very beginning of September (target date is September 1st)
2) Release a couple of alpha/beta/rc version with fixes and improvements based on the Community feedback.
3) Release a stable version sometime in October.

**Comment by MarcinZiabek** [comment]
## Document Preview Interactions

The new QuestPDF Companion application introduces more sophisticated ways to interact with your documents, offering a deeper understanding of their structure and the size of each element. This ultimately shortens the feedback loop and makes it easier to fine-tune every aspect of the document's look and feel.

The application includes four tools, each accessible with easy-to-remember shortcuts:
- Magnifier [key 1]
- Show details of the current point with magnet capability (including page number, X and Y coordinates) [key 2]
- Measure content vertically [key 3]
- Measure content horizontally [key 4]

The video below demonstrates these tools in action:

https://github.com/user-attachments/assets/a98a7582-b654-4651-b001-159d18f6eee4

**Comment by MarcinZiabek** [comment]
## Document Hierarchy and Previewer Interactions

Today, I'd like to introduce a set of new interactions between two key components of our application:

- `Document Hierarchy`: Displays the entire document structure and content in a tree view.
- `Document Preview`: Shows the document content in a rendered, visual format.


By watching the video, you will discover several important features, including but not limited to:
- `Element Selection`: Double-click anywhere in the preview to select the hovered element (typically the smallest and nearest element).
- `Focus Enhancement`: When an element is selected, all content outside of it is blurred to improve readability and focus.
- `Hover Highlights`: Hovering over elements in the Document Hierarchy tree view temporarily highlights the related content in the preview area.
- `Detailed View`: When an element is selected, additional details are displayed, such as its size, position on the screen, and key properties.
- `Multi-Page Navigation`: If an element spans multiple pages, you can navigate between these locations using arrow buttons.

https://github.com/user-attachments/assets/6d5637c3-ce92-4eac-becf-14e3527bea2d

**Comment by MarcinZiabek** [comment]
Finally! 🎊 

The first public, pre-release version of QuestPDF Companion App has been released.

1) It requires the QuestPDF library `2024.10.X` or later,
2) Introduction: https://www.questpdf.com/companion/usage.html
3) Available features: https://www.questpdf.com/companion/features.html

I target the stable release somewhere in October.

In the nearest future, once the application is stable and high-quality, its source code will be published on an identical license as the QuestPDF library.

**Reply by paladin2005** [reply]
This is awesome, I just tried and wow!!! Thanks!!! good job!

**Reply by MarcinZiabek** [reply]
Thank you 😄 

It was an extremely time-consuming project that hopefully will greatly improve the development experience. All was possible thanks to the additional funding from the new library license. So, it is also a massive thank to every company wanting to comply with our licensing terms ❤️

**Reply by rodem** [reply]
It's typo in [Usage ](https://www.questpdf.com/companion/usage.html)docs:
instead of
`using QuestPDF.Previewer;`
should be:
`using QuestPDF.Companion;`

And in [Manual font registration](https://www.questpdf.com/api-reference/text/font-management.html#manual-font-registration
) also swaped params:
instead of
`FontManager.RegisterFontWithCustomName(fontStream, "MyBarcodeFont");`
should be:
`FontManager.RegisterFontWithCustomName("MyBarcodeFont", fontStream);`

Thanks for great library!

By the way around companion.
I have dev setup:
 - Windows host with Vs Code and Companion
 - Dev via Remote SSH on Debian VM host (not WSL) on same network
 
In this setup Companion does work like in Usage help. Looking workarounds for now.
I think there must be tunel from Debian to Win host on 12500 port.
May be `ShowInCompanion(12345);` should also include IPAddress?

`fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.Exception: Cannot connect to the QuestPDF Companion tool. Please ensure that the tool is running and the port is correct. Learn more: https://www.questpdf.com/companion/usage.html
         at QuestPDF.Companion.CompanionService.CheckIfCompanionIsRunning()`

**Reply by rodem** [reply]
Setup works fine with custom tunnel tools like [rathole](https://github.com/rapiz1/rathole) or [frp](https://github.com/fatedier/frp).

**Reply by MarcinZiabek** [reply]
@rodem Indeed 😄  I have fixed both issues related to documentation. Thank you!
